### ISSUE TITLE
Themes are allowed to extend and override layouts

### ISSUE THREAD

**Issue by ericmorand** [NightlyCommit/magento2-module-soc#2]
# Violation

Magento 2 framework allows themes to extend and override layouts and forbids modules to override layouts.

# Why it is a violation

This is a violation because layouts are XML files that contains blocks classnames and blocks are module-level elements. It makes the themes dependent on data that they are not managing themselves and can't be declared as dependencies.

# Illustration of the violation

Consider the following file, taken from the [Layout instructions documentation](https://devdocs.magento.com/guides/v2.4/frontend-dev-guide/layouts/xml-instructions.html)

```xml
<block class="Magento\Catalog\Block\Product\ListProduct" name="category.products.list" as="product_list" template="Magento_Catalog::product/list.phtml"/>
```

Here, a theme declare a block of class `Magento\Catalog\Block\Product\ListProduct` under the name `category.products.list`. But it is impossible for the theme writer to know if the block class exists since blocks are declared by modules. If the block class doesn't exist in the application where the theme is installed, then the application would crash. It is impossible to predict by the theme builder.

In other words, having themes allowed to extend and override layouts makes the behavior of the final product unpredictable.

# Workarounds

Forbiding themes to extend layouts can be enforced by project policies - i.e. enforced at code-review time. Unfortunately, there doesn't seem to be a workaround to allow modules to override layouts.
